src/ircode.c

```diff
@@ -240,6 +261,16 @@ static void jl_encode_value_(jl_ircode_state *s, jl_value_t *v, int as_literal)
         write_uint8(s->s, TAG_RETURNNODE);
         jl_encode_value(s, jl_get_nth_field(v, 0));
     }
+    else if (jl_is_quotenode(v)) {
+        write_uint8(s->s, TAG_QUOTENODE);
+        jl_value_t *inner = jl_quotenode_value(v);
+        // we might need to return this exact object at run time, therefore codegen might
+        // need to reference it as well, so it is more likely useful to give it a root
+        if (jl_is_expr(inner))
```

Comment:
This should be any AST node with possible mutable state (such as PhiNode) also, right?

Reply:
PhiNodes can't be in ASTs, but also our code really shouldn't assume their identities are consistent over encoding/decoding.

Reply:
This is inside a quotenode though which should be explicitly preserving identity (egal) of the user object?

Reply:
:shrug: Ok.